cudax/include/cuda/experimental/detail/hierarchy_dimensions.cuh

```diff
@@ -0,0 +1,811 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX_DETAIL_HIERARCHY_DIMENSIONS
+#define _CUDAX_DETAIL_HIERARCHY_DIMENSIONS
+
+#include <cuda/std/tuple>
+
+#include "level_dimensions.cuh"
+#include <nv/target>
+
+namespace cuda::experimental
+{
+
+/* TODO right now operator stacking can end up with a wrong unit, we could use below type, but we would need an explicit
+ thread_level inserter
+struct unknown_unit : public hierarchy_level
+{
+  using product_type  = unsigned int;
+  using allowed_above = allowed_levels<>;
+  using allowed_below = allowed_levels<>;
+};
+*/
+
+template <typename BottomUnit, typename... Levels>
+struct hierarchy_dimensions_fragment;
+
+// If lowest unit in the hierarchy is thread, it can be considered a full hierarchy and not only a fragment
+template <typename... Levels>
+using hierarchy_dimensions = hierarchy_dimensions_fragment<thread_level, Levels...>;
+
+namespace detail
+{
+// Function to sometimes convince the compiler something is a constexpr and not really accessing runtime storage
+// Mostly a work around for what was addressed in P2280 (c++23) by leveraging a default constructor of extents
+template <typename T, size_t... Extents>
+_CCCL_HOST_DEVICE constexpr auto fool_compiler(const dimensions<T, Extents...>& ex)
+{
+  if constexpr (dimensions<T, Extents...>::rank_dynamic() == 0)
+  {
+    return dimensions<T, Extents...>();
+  }
+  else
+  {
+    return ex;
+  }
+}
+
+template <typename QueryLevel, typename Hierarchy>
+struct has_level_helper;
+
+template <typename QueryLevel, typename Unit, typename... Levels>
+struct has_level_helper<QueryLevel, hierarchy_dimensions_fragment<Unit, Levels...>>
+    : public ::cuda::std::disjunction<::cuda::std::is_same<QueryLevel, typename Levels::level_type>...>
+{};
+
+// Is this needed?
+template <typename QueryLevel, typename... Levels>
+struct has_level_helper<QueryLevel, hierarchy_dimensions<Levels...>>
+    : public ::cuda::std::disjunction<::cuda::std::is_same<QueryLevel, typename Levels::level_type>...>
+{};
+
+template <typename QueryLevel, typename Hierarchy>
+struct has_unit
+{};
+
+template <typename QueryLevel, typename Unit, typename... Levels>
+struct has_unit<QueryLevel, hierarchy_dimensions_fragment<Unit, Levels...>> : ::cuda::std::is_same<QueryLevel, Unit>
+{};
+
+template <unsigned int Id, typename... Levels>
+using level_at_index = typename ::cuda::std::tuple_element<Id, ::cuda::std::tuple<Levels...>>::type;
```

Comment:
If we only want the access at the Nth element we should rather use the simple `__tuple_types``rather than the full fledged `tuple`